Use slf4j adapters for logging.

mfriedenhagen opened the pull request:
- By introducing adapters for log4j and commons-logging and
- including the original log4j and commons-logging dependencies as provided
  all logging should be redirected to slf4j and finally java.util.logging.

This is an enhancement to #1765

Commit:
Use slf4j adapters for logging.

- By introducing adapters for log4j and commons-logging and
- including the original log4j and commons-logging dependencies as provided
all logging should be redirected to slf4j and finally java.util.logging.

Commit:
Include slf4j-jdk14 during tests.

jtnord commented on pom.xml:
```diff
@@ -215,14 +215,21 @@ THE SOFTWARE.
+        <groupId>commons-logging</groupId>
+        <artifactId>commons-logging</artifactId>
+        <version>1.1.3</version>
+        <scope>provided</scope><!-- by jcl-over-slf4j -->
```
this is wrong and shouldn't be included

mfriedenhagen replied:
@jtnord, we have used this technique successfully for years in our department pom. By specifying commons-logging as provided, you do not have to exclude it in every other current dependency *and* inhibit that it creeps in by new versions/dependencies which might introduce commons-logging as transitive dependency.  One could argue about the location, though: it would even be better to include this is in `org.jenkins-ci:jenkins`.

jtnord replied:
it can cause havoc with tests though :-(
perhaps `<scope>none</scope>` if that ~~hack~~ undocumented feature still works with the latest maven versions.

jtnord commented on pom.xml:
```diff
@@ -664,7 +671,6 @@ THE SOFTWARE.
                 <bannedDependencies>
                   <excludes>
                     <exclude>org.sonatype.sisu:sisu-guice</exclude>
-                    <exclude>log4j:log4j</exclude>
```
this should be here as well as the banning of commons-logging.
otherwise plugins can bundle the library by overriding the scope - which they may already have done (which is just a warning - rather than a build breakage)

mfriedenhagen replied:
OK, hopefully plugins which override the scope on purpose (and changed it to compile or runtime, test does not matter) might ignore the warning as well, but a warning might be a good idea nontheless. Then I would include `commons-logging:commons-logging` as well?

jtnord replied:
yes - and these don't produce warnings but errors and fail the build.

mfriedenhagen replied:
I now modified the enforcer-rules to limit the exclusions to scopes compile and runtime.

jtnord commented on pom.xml:
```diff
@@ -215,14 +215,21 @@ THE SOFTWARE.
-        <version>1.1.3</version>
+        <groupId>log4j</groupId>
+        <artifactId>log4j</artifactId>
+        <version>1.2.17</version>
```
ditto

mfriedenhagen replied:
ditto ;-)

Commit:
Reintroduce enforcer rules as suggested by @jtnord.

Commit:
Only exclude commons-logging and log4j in dangerous scopes compile and runtime.

Commit:
Exclude commons-logging and log4j in dangerous scopes compile and runtime in plugins as well.

jtnord commented on plugins/pom.xml:
```diff
@@ -80,6 +82,32 @@
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>${slf4jVersion}</version>
+      <scope>provided</scope>
```
dependencyManagement rather than dependencies ?

mfriedenhagen replied:
I thought it would be a good idea to include this as a dependency, so people know, that they should use slf4j for their own logging purposes. As it is marked as provided, it will not be included in the HPI but is available during compile and test.

TheMint commented:
hm... this conflict with my idea... but is a good idea... best than mine :+1:

Commit:
Manage commons-logging and log4j as provided, include slf4j adapters for tests.

 By means of this *all* logging should go to JUL.  By defining slf4j-api as compile
 scope we mark it as the one to be used, while by declaring it optional we make
 sure it is not packaged in the resulting HPI.

mfriedenhagen commented:
I tested 1003796ff22ee by something like this in a using plugin:

```java
	@Extension
	public static class DescriptorImpl extends Descriptor<RepositoryBrowser<?>> {
		static org.slf4j.Logger SLF4J = org.slf4j.LoggerFactory.getLogger(DescriptorImpl.class);
		static org.apache.commons.logging.Log JCL = org.apache.commons.logging.LogFactory.getLog(DescriptorImpl.class);
		static org.apache.log4j.Logger LOG4J = org.apache.log4j.Logger.getLogger(DescriptorImpl.class);
		static java.util.logging.Logger JUL = java.util.logging.Logger.getLogger(DescriptorImpl.class.getName());
		{
			SLF4J.info("SLF4J");
			JCL.info("JCL");
			LOG4J.info("LOG4J");
			JUL.info("JUL");
		}
```

* When running `mvn hpi:run`, everything was logged correctly via JUL
* No logging libraries were incidentally packaged in the HPI. 
* When I included `commons-logging` or `log4j` in scopes `compile` or `runtime`, enforcer broke the build as intended.

jtnord commented:
:+1:

mfriedenhagen commented:
Not sure, why the build broke. Unfortunately, I am not allowed to trigger
it again.

mfriedenhagen commented:
I ran the test locally in Intellij and it did succeed.
Regards Mirko
--
http://illegalstateexception.blogspot.com/
https://github.com/mfriedenhagen/ (http://osrc.dfm.io/mfriedenhagen)
https://bitbucket.org/mfriedenhagen/


On Thu, Sep 3, 2015 at 7:53 PM, Mirko Friedenhagen
<mfriedenhagen@gmail.com> wrote:
> Not sure, why the build broke. Unfortunately, I am not allowed to trigger it
> again.

jtnord commented:
re-triggered build works

jtnord commented:
Thanks Mirko.

mfriedenhagen commented:
Hello James, you're welcome. I wonder whether we should create a JIRA issue
to document this change or at least mention it in the changelog because
there might be hidden problems with some plugins.

jtnord commented:
noted in 93dfe3377ec8d430818f5b9073f16c677343adb4

mfriedenhagen commented:
Thank you, James.
